File: pairboarding/day1.rb
def digital_root(num)
  until num < 10
    digit_sum = 0
    partial_num = num
    until partial_num == 0
      digit_sum += partial_num % 10
      partial_num /= 10
    end
    num = digit_sum
  end
  num
end

def caesar_cipher(string, shift)
  string.split('').map do |char|
    if char.ord.between?("a".ord, "a".ord + 25)
      ((((char.ord + shift) - "a".ord) % 26) + "a".ord).chr
    else
      char
    end
  end.join('')
end

def common_substrings(str1, str2)
  longest = ""
  str1.length.times do |start_idx|
    end_idx = start_idx + longest.length
    while end_idx < str1.length
      longest = str1[start_idx..end_idx] if str2.include?(str1[start_idx..end_idx])
      end_idx += 1
    end
  end
  longest
end

def sum_rec(array)
  return 0 if array.empty?
  array.first + sum_rec(array.drop(1))
end

def fibs(num)
  fibs = [0, 1]
  return fibs.take(num) if num <= 2
  until fibs.length == num
    fibs.push(fibs[-1] + fibs[-2])
  end
  fibs
end

def is_palindrome?(string)
  i, j = 0, string.length - 1
  while i < j
    i += 1 until string[i] != " "
    j -= 1 until string[j] != " "
    return false if string[i] != string[j]
    i += 1
    j -= 1
  end
  true
end

File: pairboarding/day1.js
function digitalRoot(num) {
  while (num >= 10) {
    var sum = 0;
    var temp = num;
    while (temp > 0) {
      var remainder = temp % 10;
      sum += remainder;
      temp = (temp - remainder) / 10;
    }
    num = sum;
  }
  return num;
}

function caesarCipher(string, shift) {
  var alpha = "abcdefghijklmnopqrstuvwxyz";
  var new_string = "";
  for (i = 0; i < string.length; i++) {
    if (string[i] === " ") {
      new_string += " ";
    } else {
      var old_idx = alpha.split("").indexOf(string[i]);
      var new_idx = (old_idx + shift) % alpha.length;
      new_string += alpha[new_idx];
    }
  }
  return new_string;
}

function commonSubstrings(str1, str2) {
  var longest = "";
  for (startIdx = 0; startIdx < str1.length; startIdx++) {
    for (diff = longest.length + 1; startIdx + diff < str1.length + 1; diff++) {
      var substring = str1.slice(startIdx, startIdx + diff);
      var start2 = 0;
      while (start2 < str2.length - diff + 1) {
        if (str2.slice(start2, start2 + diff) === substring) {
          longest = substring;
        }
        start2 += 1;
      }
    }
  }
  return longest;
}

function sum_rec(nums) {
  if (nums.length === 0) {
    return 0;
  }
  return nums[0] + sum_rec(nums.slice(1));
}
